webui/react/src/utils/analytics.ts

```diff
@@ -0,0 +1,75 @@
+import { DeterminedInfo } from 'types';
+
+interface InternalSegmentAnalytics extends SegmentAnalytics.AnalyticsJS {
+  methods: string[];
+}
+
+interface AnalyticsData {
+  analytics?: SegmentAnalytics.AnalyticsJS;
+  isEnabled: boolean;
+}
+
+/*
+ * This object is stored outside of the React space to allow analytics to be
+ * accessible outside of React components, such as `ErrorHandler`.
+ */
+const data: AnalyticsData = {
+  analytics: undefined,
+  isEnabled: false,
+};
+
+/*
+ * Segment analytics library is loaded on `index.html` dynamically through a JS
+ * snippet. The library has a `methods` string array which is an inventory of
+ * methods to load. Though there is a `ready` callback, it is not reliably called
+ * and it is lacking a boolean ready state. Instead when all the methods are defined
+ * on the library object `window.analytics` it is considered ready.
+ */
+const getReadyAnalytics = (): SegmentAnalytics.AnalyticsJS | undefined => {
+  if (data.analytics) return data.analytics;
+
+  const analytics: InternalSegmentAnalytics = window.analytics;
+  if (analytics) {
+    const keys = Object.keys(analytics).reduce((acc, key) => {
+      acc[key] = true;
+      return acc;
+    }, {} as Record<string, boolean>);
+
+    const methods: string[] = analytics.methods || [];
+    const missingMethods = methods.some(method => keys[method] === undefined);
+    if (!missingMethods) {
+      data.analytics = analytics;
+      return data.analytics;
+    }
+  }
+  return undefined;
+};
+
+export const setupAnalytics = (info: DeterminedInfo): void => {
+  if (!data.analytics) data.analytics = getReadyAnalytics();
+  if (!data.analytics || data.isEnabled) return;
+
+  /*
+   * Segment key should be 32 characters composed of upper case letters,
+   * lower case letters and numbers 0-9.
+   */
+  const isEnabled = info.telemetry.enabled;
+  const isValidKey = info.telemetry.segmentKey && /^[a-z0-9]{32}$/i.test(info.telemetry.segmentKey);
+  if (isEnabled && isValidKey) {
+    data.analytics.load(info.telemetry.segmentKey || '');
```

Comment:
question: why would this `load` be always available compared to how it was before? is it because we trigger it in a hook in App.tsx ?

blob:https://determinedai.atlassian.net/0cbb1431-44c7-44af-bf4a-8728c3f0da5b#media-blob-url=true&id=b6131886-6e8e-443b-beb5-6f4bce812d59&contextId=13662&collection=

Reply:
ah good question. It's because it's defined in the `index.html` snippet.

Reply:
update: the link should be to the image here https://determinedai.atlassian.net/browse/DET-3636